File: shoring/desk/src/main/java/com/panopset/marin/games/blackjack/BlackjackGameController.java
package com.panopset.marin.games.blackjack;

import java.net.URL;
import java.util.ResourceBundle;

import com.panopset.blackjackEngine.BlackjackGameEngine;
import com.panopset.blackjackEngine.CycleSnapshot;
import com.panopset.fxapp.FontManagerFX;
import com.panopset.fxapp.ReflectorFX;

import javafx.animation.AnimationTimer;
import javafx.application.Platform;
import javafx.fxml.FXML;
import javafx.fxml.Initializable;
import javafx.scene.canvas.Canvas;
import javafx.scene.canvas.GraphicsContext;
import javafx.scene.input.KeyEvent;
import javafx.scene.layout.FlowPane;
import javafx.scene.paint.Color;

public class BlackjackGameController implements Initializable {

  public void initialize(URL location, ResourceBundle resources) {
    ReflectorFX.bindBundle(this, resources);
    startPaintCycle();
  }

  private void handleKey(KeyEvent keyEvent) {
    if (bge == null) {
      return;
    }
    bge.exec(keyEvent.getText().toLowerCase());
  }

  private int fontSize = 0;
  private boolean gameStarted = false;
  private CycleSnapshot paintedSnapshot;
  private boolean dirty = true;
  private AnimationTimer timer;

  public AnimationTimer getTimer() {
    if (timer == null) {
      timer = new AnimationTimer() {
        @Override
        public void handle(long now) {
          if (now - lastUpdate > 5000000L) {
            paintFelt();
            lastUpdate = now;
          }
        }
      };

    }
    return timer;
  }

  long lastUpdate = 0;

  private void startPaintCycle() {
    Platform.runLater(() -> {
      getTimer().start();
    });
  }

  private boolean isDirty() {
    if (!dirty) {
      if (bge == null) {
        return false;
      }
      if (bge.getCurrentSnapshot() == null) {
        return false;
      }
      if (paintedSnapshot != bge.getCurrentSnapshot()) {
        paintedSnapshot = bge.getCurrentSnapshot();
        dirty = true;
      }
      if (fontSize != FontManagerFX.getSize()) {
        fontSize = FontManagerFX.getSize();
        dirty = true;
      }
    }
    return dirty;
  }

  public void update() {
    dirty = true;
  }

  private CycleSnapshot paintFelt() {
    if (felt == null) {
      return null;
    }
    int layoutHeight = (int) felt.getParent().getLayoutBounds().getHeight();
    int layoutWidth = (int) felt.getParent().getLayoutBounds().getWidth();

    if (layoutHeight < 10 || layoutWidth < 10) {
      return null;
    }

    felt.setWidth(layoutWidth);
    felt.setHeight(layoutHeight);

    GraphicsContext g = felt.getGraphicsContext2D();


    if (fxmlGlue == null) {
      error(g, "fxmlGlue is null.");
      return null;
    }

    if (!fxmlGlue.isSet()) {
      g.setFill(Color.BLANCHEDALMOND);
      g.fillRect(0, 0, layoutWidth, layoutHeight);
      g.setFill(Color.DARKGREEN);
      g.fillText(fxmlGlue.toString(), 100, 100);
      return null;
    }

    if (getEngine() == null) {
      error(g, "bge is null.");
      return null;
    }

    if (!bge.isActive()) {
      g.setFill(Color.DARKRED);
      g.fillRect(0, 0, layoutWidth, layoutHeight);
      return null;
    }

    CycleSnapshot rtn = bge.getCurrentSnapshot();

    if (!gameStarted) {
      felt.getScene().setOnKeyPressed(this::handleKey);
      gameStarted = true;
    }

    if (!isDirty()) {
      return null;
    }
    getFeltPainter().draw(rtn, g, layoutWidth, layoutHeight);



    dirty = false;

    return rtn;
  }

  private void error(GraphicsContext g, String msg) {
    g.setFill(Color.DARKRED);
    g.fillRect(0, 0, 1000, 1000);
    g.setFill(Color.YELLOW);
    g.fillText(msg, 100, 100);
  }

  private FxmlGlue fxmlGlue;

  public void setFxmlGlue(FxmlGlue fxmlGlue) {
    this.fxmlGlue = fxmlGlue;
  }

  private FeltPainter fp;

  private FeltPainter getFeltPainter() {
    if (fp == null) {
      fp = new FeltPainter();
    }
    return fp;
  }

  @FXML
  public Canvas felt;

  @FXML
  public FlowPane gameControlFlow;

  private BlackjackGameEngine bge;

  public void setEngine(BlackjackGameEngine bge) {
    this.bge = bge;
  }

  private BlackjackGameEngine getEngine() {
    if (bge == null) {
      if (fxmlGlue.isSet()) {
        bge = new BlackjackGameEngineFactory().create(fxmlGlue);
        bge.getZombie().addStopAction(getTimer()::stop);
      }
    }
    return bge;
  }
}

File: code/shoring/desk/src/main/kotlin/com/panopset/marin/games/blackjack/BlackjackGameController.kt
package com.panopset.marin.games.blackjack

import com.panopset.blackjackEngine.BlackjackGameEngine
import com.panopset.blackjackEngine.CMD_AUTO
import com.panopset.blackjackEngine.CMD_RESET
import com.panopset.blackjackEngine.BlackjackGameState
import com.panopset.compat.Zombie
import com.panopset.desk.games.bj.BlackjackFxControls
import com.panopset.fxapp.FontManagerFX
import com.panopset.fxapp.FxDoc
import javafx.animation.AnimationTimer
import javafx.application.Platform
import javafx.event.EventHandler
import javafx.scene.input.KeyEvent
import javafx.scene.paint.Color
import java.util.*
import kotlin.concurrent.thread

class BlackjackGameController(ctls: BlackjackFxControls, val bge: BlackjackGameEngine) {
    val fxDoc: FxDoc = ctls.fxDoc
    var felt = ctls.felt
    private var dirty = false
    private var firstTime = true

    init {
        Zombie.addStopAction { timer.stop() }
        if (bge.bankroll.reloadAmount == 0) {
            bge.frontEndPreInitCheck()
        }
        startPaintCycle()
    }

    private fun handleKey(keyEvent: KeyEvent) {
        val action = keyEvent.text.lowercase(Locale.getDefault())
        if (bge.isAutomaticRunning()) {
            if (action != CMD_AUTO) {
                bge.exec(CMD_AUTO)
                return
            }
        }
        bge.exec(action)
    }

    private var fontSize = 0
    private var paintedSnapshot: BlackjackGameState? = null
    private var timer: AnimationTimer = object : AnimationTimer() {
                    override fun handle(now: Long) {
                        if (now - lastUpdate > 5000000L) {
                            paintFelt()
                            lastUpdate = now
                        }
                    }
                }

    var lastUpdate: Long = 0
    private var oneSecond: Long = 0

    private fun startPaintCycle() {
        Platform.runLater { timer.start() }
    }

    private fun isDirty(): Boolean {
        if (fontSize != FontManagerFX.getSize()) {
            fontSize = FontManagerFX.getSize()
            dirty = true
        }
        if (!dirty) {
            val currentTime = Date().time
            if (currentTime - oneSecond > 1000) {
                oneSecond = currentTime
                dirty = true
            }
        }
        return dirty
    }

    fun update() {
        dirty = true
    }

    private var binding = false
    //var dbgcount = 0L

    private fun paintFelt() {
        if (binding) {
            return
        }
        if (felt.parent == null) {
            return
        }
        val layoutHeight = felt.parent.layoutBounds.height.toInt()
        val layoutWidth = felt.parent.layoutBounds.width.toInt()
        if (layoutHeight < 10 || layoutWidth < 10) {
            return
        }
        felt.width = layoutWidth.toDouble()
        felt.height = layoutHeight.toDouble()
        val g = felt.graphicsContext2D

//            g.fill = Color.BLANCHEDALMOND
//            g.fillRect(0.0, 0.0, layoutWidth.toDouble(), layoutHeight.toDouble())
//            g.fill = Color.DARKGREEN
//            g.fillText("diags ${dbgcount++}", 100.0, 100.0)

        if (!Zombie.isActive) {
            g.fill = Color.DARKRED
            g.fillRect(0.0, 0.0, layoutWidth.toDouble(), layoutHeight.toDouble())
            return
        }
        if (isDirty()) {
            if (firstTime) {
                firstTime = false
                felt.scene.onKeyPressed = EventHandler {
                        keyEvent: KeyEvent -> handleKey(keyEvent)
                }
                thread {
                    bge.exec(CMD_RESET)
                }
            }
        } else {
            return
        }
        FeltPainter().draw(fxDoc, bge.getLatestSnapshot(), g, layoutWidth, layoutHeight)
        dirty = false
        return
    }
}
